### PR TITLE
fix: apply decimal place rounding

### PR DESCRIPTION
**Pull request by ironAiken2**
### TL;DR

Handle visible decimal points according to [backend.ai writing rules](https://loop.cloud.microsoft/p/eyJ3Ijp7InUiOiJodHRwczovL2xhYmx1cC5zaGFyZXBvaW50LmNvbS8%2FbmF2PWN6MGxNa1ltWkQxaUlXZG1ZbUpTWDA1eFMxVnhkblpoY0hrd2MyeDFaV05ZTnkxVmJ6TlRXVkpCZFdod1dFNUJiMGhxTFVOWVRqRjJSRVZrY3poVWNFeEpNR1JvVWt0bWRuTW1aajB3TVVkTlQwaEpVVk5WUjBOYVJGQllOa3hTTlVKWlYwNVVOMUJSVjA1S1F6ZFhKbU05Sm1ac2RXbGtQVEUlM0QiLCJyIjpmYWxzZX0sInAiOnsidSI6Imh0dHBzOi8vbGFibHVwLnNoYXJlcG9pbnQuY29tL2NvbnRlbnRzdG9yYWdlL0NTUF80N2RiZjY4MS02YWYzLTRhMjktYWZiZC1hYTcyZDJjOTZlNzk%2FbmF2PWN6MGxNa1pqYjI1MFpXNTBjM1J2Y21GblpTVXlSa05UVUY4ME4yUmlaalk0TVMwMllXWXpMVFJoTWprdFlXWmlaQzFoWVRjeVpESmpPVFpsTnprbVpEMWlJV2RtWW1KU1gwNXhTMVZ4ZG5aaGNIa3djMngxWldOWU55MVZiek5UV1ZKQmRXaHdXRTVCYjBocUxVTllUakYyUkVWa2N6aFVjRXhKTUdSb1VrdG1kbk1tWmowd01VZE5UMGhKVVZKU1RsRk1Va2hEVWxkWlJrSktNa3N5TkVkUFFWcFlWbG8wSm1NOUptWnNkV2xrUFRFJTNEIiwiciI6ZmFsc2V9LCJpIjp7ImkiOiI5NTM1ZjRmMy1iMTE4LTQwZDQtOWVhNS05OTliMDJlYjdiYjkifX0%3D)

### Changes 
- Allow string as input type for the `toFixedFloorWithoutTrailingZeros` function. Changed logic to handle exception cases.
- add new `toFixedWithTypeValidation` function to show zero-terminated decimals as they are (e.g. 1.00, 2.10 for GPU or fGPU)
- The same functionality has been applied to webcomponent.

### How to Test
- Verify that decimalization is properly applied to the `resource-panel` and `resource-monitor` parts of the summary page.
- Verify that the decimal points shown on the Agent tab of the Resource Page are handled correctly.

**Checklist:** (if applicable)

- [ ] Mention to the original issue
- [ ] Documentation
- [ ] Minium required manager version
- [ ] Specific setting for review (eg., KB link, endpoint or how to setup)
- [ ] Minimum requirements to check during review
- [ ] Test case(s) to demonstrate the difference of before/after

### CODE REVIEW

**Comment by ironAiken2** [comment]
* **#2518** <a href="https://app.graphite.dev/github/pr/lablup/backend.ai-webui/2518?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ironAiken2 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/components/backend-ai-resource-monitor.ts [hunk 1]
```diff
@@ -643,25 +649,41 @@ export default class BackendAiResourceMonitor extends BackendAIPage {
...
                   100.0}"
-                  description="${this.used_resource_group_slot.cpu}/${this
-                    .total_resource_group_slot.cpu}"
+                  description="${this._prefixFormat(
+                    this.used_resource_group_slot.cpu,
+                    1,
+                  )} / ${this._prefixFormat(
+                    this.total_resource_group_slot.cpu,
+                    1,
+                  )} ${_t('summary.Cores')}"
```
Comment (agatha197) [inline src/components/backend-ai-resource-monitor.ts]: I'm just curious. Why did you add a translation only for this?

Comment (ironAiken2) [reply src/components/backend-ai-resource-monitor.ts]: I was going to add it because the webui doesn't provide a translation for the units, but then I realized that it's better to unify the units in English. I think it's better to remove it.

File: react/src/helper/index.tsx [hunk 2]
```diff
@@ -373,3 +373,21 @@ export const generateRandomString = (n = 3) => {
...
+ * If there is no number, it will return '0' to use as a default value when the number isn't set.
+ * @param {number} prefix - prefix to format
+ * @returns {string}
+ * formatted number with prefix. If the number is a Integer, it will return the number as a string without a decimal point.
+ */
+export const prefixFormat = (number: any, prefix: number) => {
+  if (isNaN(number)) return '0';
+  const parsedNumber = typeof number === 'string' ? parseFloat(number) : number;
+  const parsedNum =
+    typeof parsedNumber === 'string' ? parseFloat(parsedNumber) : parsedNumber;
```
Comment (agatha197) [inline react/src/helper/index.tsx]: these lines are duplicated.

File: src/components/backend-ai-resource-monitor.ts [hunk 3]
```diff
@@ -1015,7 +1145,7 @@ export default class BackendAiResourceMonitor extends BackendAIPage {
                       "
                       >
                         ${this._numberWithPostfix(
-                          this.used_slot_percent.atom_plus_device,
+                          (this.used_slot_percent.atom_plus_device, 1),
```
Comment (agatha197) [inline src/components/backend-ai-resource-monitor.ts]: Doesn't it need `this._prefixFormat`?

**Review by agatha197 — CHANGES_REQUESTED** [review]
From the [backend.ai writing rules](https://loop.cloud.microsoft/p/eyJ3Ijp7InUiOiJodHRwczovL2xhYmx1cC5zaGFyZXBvaW50LmNvbS8%2FbmF2PWN6MGxNa1ltWkQxaUlXZG1ZbUpTWDA1eFMxVnhkblpoY0hrd2MyeDFaV05ZTnkxVmJ6TlRXVkpCZFdod1dFNUJiMGhxTFVOWVRqRjJSRVZrY3poVWNFeEpNR1JvVWt0bWRuTW1aajB3TVVkTlQwaEpVVk5WUjBOYVJGQllOa3hTTlVKWlYwNVVOMUJSVjA1S1F6ZFhKbU05Sm1ac2RXbGtQVEUlM0QiLCJyIjpmYWxzZX0sInAiOnsidSI6Imh0dHBzOi8vbGFibHVwLnNoYXJlcG9pbnQuY29tL2NvbnRlbnRzdG9yYWdlL0NTUF80N2RiZjY4MS02YWYzLTRhMjktYWZiZC1hYTcyZDJjOTZlNzk%2FbmF2PWN6MGxNa1pqYjI1MFpXNTBjM1J2Y21GblpTVXlSa05UVUY4ME4yUmlaalk0TVMwMllXWXpMVFJoTWprdFlXWmlaQzFoWVRjeVpESmpPVFpsTnprbVpEMWlJV2RtWW1KU1gwNXhTMVZ4ZG5aaGNIa3djMngxWldOWU55MVZiek5UV1ZKQmRXaHdXRTVCYjBocUxVTllUakYyUkVWa2N6aFVjRXhKTUdSb1VrdG1kbk1tWmowd01VZE5UMGhKVVZKU1RsRk1Va2hEVWxkWlJrSktNa3N5TkVkUFFWcFlWbG8wSm1NOUptWnNkV2xrUFRFJTNEIiwiciI6ZmFsc2V9LCJpIjp7ImkiOiI5NTM1ZjRmMy1iMTE4LTQwZDQtOWVhNS05OTliMDJlYjdiYjkifX0%3D), let's modify the fixed unit. For example, the unit/capacity of RAM is expressed up to the second decimal place.

**Comment by ironAiken2** [comment]
> From the [backend.ai writing rules](https://loop.cloud.microsoft/p/eyJ3Ijp7InUiOiJodHRwczovL2xhYmx1cC5zaGFyZXBvaW50LmNvbS8%2FbmF2PWN6MGxNa1ltWkQxaUlXZG1ZbUpTWDA1eFMxVnhkblpoY0hrd2MyeDFaV05ZTnkxVmJ6TlRXVkpCZFdod1dFNUJiMGhxTFVOWVRqRjJSRVZrY3poVWNFeEpNR1JvVWt0bWRuTW1aajB3TVVkTlQwaEpVVk5WUjBOYVJGQllOa3hTTlVKWlYwNVVOMUJSVjA1S1F6ZFhKbU05Sm1ac2RXbGtQVEUlM0QiLCJyIjpmYWxzZX0sInAiOnsidSI6Imh0dHBzOi8vbGFibHVwLnNoYXJlcG9pbnQuY29tL2NvbnRlbnRzdG9yYWdlL0NTUF80N2RiZjY4MS02YWYzLTRhMjktYWZiZC1hYTcyZDJjOTZlNzk%2FbmF2PWN6MGxNa1pqYjI1MFpXNTBjM1J2Y21GblpTVXlSa05UVUY4ME4yUmlaalk0TVMwMllXWXpMVFJoTWprdFlXWmlaQzFoWVRjeVpESmpPVFpsTnprbVpEMWlJV2RtWW1KU1gwNXhTMVZ4ZG5aaGNIa3djMngxWldOWU55MVZiek5UV1ZKQmRXaHdXRTVCYjBocUxVTllUakYyUkVWa2N6aFVjRXhKTUdSb1VrdG1kbk1tWmowd01VZE5UMGhKVVZKU1RsRk1Va2hEVWxkWlJrSktNa3N5TkVkUFFWcFlWbG8wSm1NOUptWnNkV2xrUFRFJTNEIiwiciI6ZmFsc2V9LCJpIjp7ImkiOiI5NTM1ZjRmMy1iMTE4LTQwZDQtOWVhNS05OTliMDJlYjdiYjkifX0%3D), let's modify the fixed unit. For example, the unit/capacity of RAM is expressed up to the second decimal place.

Applied! 🙂

File: src/components/backend-ai-resource-monitor.ts [hunk 4]
```diff
@@ -1129,16 +1311,24 @@ export default class BackendAiResourceMonitor extends BackendAIPage {
...
-                    .concurrency_max === 1000000
+                  description="${this._prefixFormatWithTrailingZeros(
+                    this.concurrency_used,
+                    0,
+                  )} / ${this.concurrency_max === 1000000
                     ? '∞'
-                    : parseInt(this.concurrency_max)}"
+                    : this._prefixFormatWithTrailingZeros(
+                        this.concurrency_max,
+                        1,
```
Comment (agatha197) [inline src/components/backend-ai-resource-monitor.ts]:
```suggestion
                        2,
```

**Review by agatha197 — CHANGES_REQUESTED** [review]
image
Can you widen progress bar too?

**Comment by ironAiken2** [comment]
> image
> Can you widen progress bar too?

On the sesson list page, when backend-ai-resource-monitor is used, the progress bar width was changing to narrow, I changed it to keep the existing width. 

I think that would require a design change to handle that. The resource monitor part will be getting a redesign soon, so why not use a different stack to handle it? @agatha197

**Comment by agatha197** [comment]
> > image
> > Can you widen progress bar too?
> 
> On the sesson list page, when backend-ai-resource-monitor is used, the progress bar width was changing to narrow, I changed it to keep the existing width.
> 
> I think that would require a design change to handle that. The resource monitor part will be getting a redesign soon, so why not use a different stack to handle it? kimsujin (@agatha197)

hmm.. okay.

**Review by agatha197 — APPROVED** [review]
LGTM
